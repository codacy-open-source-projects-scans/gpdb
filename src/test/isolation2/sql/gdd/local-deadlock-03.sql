
-- gdd can also detect local deadlocks, however it might break at
-- different node with the local deadlock detector.  To make the local
-- deadlock testcases stable we reset the gdd period to 2min so should
-- not be triggered during the local deadlock tests.
ALTER SYSTEM SET gp_global_deadlock_detector_period to '2min';
SELECT pg_reload_conf();
-- start new session, which should always have newly reflected value
1: SHOW gp_global_deadlock_detector_period;

DROP TABLE IF EXISTS t03;
CREATE TABLE t03 (id int, val int);
INSERT INTO t03 (id, val) SELECT i, i FROM generate_series(1, 100) i;

-- gang creation order is important, reset any guc to force the creation
10: RESET optimizer;
20: RESET optimizer;
30: RESET optimizer;

10: BEGIN;
20: BEGIN;
30: BEGIN;

10: UPDATE t03 SET val=val WHERE id IN (segid(0,2), segid(1,2));

20: UPDATE t03 SET val=val WHERE id=segid(0,1);

30: UPDATE t03 SET val=val WHERE id=segid(1,1);

-- seg 0: con10 ==> con20, xid lock
-- seg 1: con10 ==> con30, xid lock
10&: UPDATE t03 SET val=val WHERE id IN (segid(0,1), segid(1,1));

-- seg 1: con30 ==> con10, xid lock
30>: UPDATE t03 SET val=val WHERE id=segid(1,2);

-- a local deadlock between con10 & con30 is formed on seg1,
-- con10 will be cancelled by local deadlock detector.
-- note, we must join c20 before c10, otherwise the test will run into infinite
-- waiting if the local deadlock detector cancels c30 instead of c10.
30<:
30q:
20<:
20q:
10<:
10q:
