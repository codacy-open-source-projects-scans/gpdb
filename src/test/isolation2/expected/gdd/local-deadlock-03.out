
-- gdd can also detect local deadlocks, however it might break at
-- different node with the local deadlock detector.  To make the local
-- deadlock testcases stable we reset the gdd period to 2min so should
-- not be triggered during the local deadlock tests.
ALTER SYSTEM SET gp_global_deadlock_detector_period to '2min';
ALTER SYSTEM
SELECT pg_reload_conf();
 pg_reload_conf 
----------------
 t              
(1 row)
-- start new session, which should always have newly reflected value
1: SHOW gp_global_deadlock_detector_period;
 gp_global_deadlock_detector_period 
------------------------------------
 2min                               
(1 row)

DROP TABLE IF EXISTS t03;
DROP TABLE
CREATE TABLE t03 (id int, val int);
CREATE TABLE
INSERT INTO t03 (id, val) SELECT i, i FROM generate_series(1, 100) i;
INSERT 0 100

-- gang creation order is important, reset any guc to force the creation
10: RESET optimizer;
RESET
20: RESET optimizer;
RESET
30: RESET optimizer;
RESET

10: BEGIN;
BEGIN
20: BEGIN;
BEGIN
30: BEGIN;
BEGIN

10: UPDATE t03 SET val=val WHERE id IN (segid(0,2), segid(1,2));
UPDATE 2

20: UPDATE t03 SET val=val WHERE id=segid(0,1);
UPDATE 1

30: UPDATE t03 SET val=val WHERE id=segid(1,1);
UPDATE 1

-- seg 0: con10 ==> con20, xid lock
-- seg 1: con10 ==> con30, xid lock
10&: UPDATE t03 SET val=val WHERE id IN (segid(0,1), segid(1,1));  <waiting ...>

-- seg 1: con30 ==> con10, xid lock
30>: UPDATE t03 SET val=val WHERE id=segid(1,2);  <waiting ...>

-- a local deadlock between con10 & con30 is formed on seg1,
-- con10 will be cancelled by local deadlock detector.
-- note, we must join c20 before c10, otherwise the test will run into infinite
-- waiting if the local deadlock detector cancels c30 instead of c10.
30<:  <... completed>
UPDATE 1
30q: ... <quitting>
20<:  <... completed>
FAILED:  Execution failed
20q: ... <quitting>
10<:  <... completed>
ERROR:  deadlock detected  (seg1 127.0.1.1:7003 pid=52248)
DETAIL:  Process 52248 waits for ShareLock on transaction 632; blocked by process 52265.
Process 52265 waits for ShareLock on transaction 631; blocked by process 52248.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (0,1) in relation "t03"
10q: ... <quitting>
